### ISSUE TITLE
conv2d on Metal silently returns all-zero output for large inputs (channel-count dependent, ~2e9 elements)

### ISSUE THREAD

**Issue by kayadibi1** [ml-explore/mlx#3979]
## Summary

`mlx.nn.Conv2d` / `mx.conv2d` on Metal **silently returns an all-zero tensor** for
large batched inputs. No exception, no warning — the op completes and produces
zeros. The failure depends on both total size and channel count, which suggests a
size-limited kernel variant selected for certain channel configurations.

This is distinct from (and much more dangerous than) the loud
`metal::malloc ... maximum allowed buffer size` error for >30 GB buffers: that one
fails the run; this one corrupts results silently.

## Environment

- mlx **0.32.0** (via pip)
- Apple **M5 Pro, 48 GB**, macOS **27.0** (build 26A5378n)
- Python 3.14.6

## Minimal repro

```python
import mlx.core as mx
import mlx.nn as nn
import numpy as np

def make(BT, H, W, C):  # deterministic non-constant input
    bt = (mx.arange(BT).reshape(BT, 1, 1, 1) % 13).astype(mx.float32) * 0.31
    h  = (mx.arange(H).reshape(1, H, 1, 1) % 7).astype(mx.float32) * 0.17
    w  = (mx.arange(W).reshape(1, 1, W, 1) % 5).astype(mx.float32) * 0.11
    c  = (mx.arange(C).reshape(1, 1, 1, C) % 3).astype(mx.float32) * 0.07
    x = mx.broadcast_to(bt + h + w + c - 1.0, (BT, H, W, C)).astype(mx.float32)
    mx.eval(x)
    return x

def test(BT, H, W, C):
    x = make(BT, H, W, C)
    conv = nn.Conv2d(C, C // 2, 3, padding=1)
    mx.eval(conv.parameters())
    y = conv(x); mx.eval(y)
    for b in (0, BT - 1):
        got  = np.array(y[b, :4, :4])
        want = np.array(conv(x[b : b + 1])[0, :4, :4])   # same op, single slice
        d = float(np.abs(got - want).max())
        print(f"BT={BT} C={C} elems={x.size:,}: batch {b} "
              f"maxdiff={d:.4f} out_std={got.std():.4f}"
              f"{'   <-- ALL ZEROS' if got.std() < 1e-6 else ''}")

test(32, 512, 512, 192)   # 1.61e9 elements -> correct
test(40, 512, 512, 192)   # 2.01e9 elements -> ALL ZEROS, silently
test(96, 512, 512, 128)   # 3.22e9 elements -> correct (!)
```

## Measured boundary (all fp32, H=W=512, kernel 3x3, padding 1)

| C_in | input elements | result |
|---|---|---|
| 192 | 1.61e9 (BT=32) | correct |
| 192 | **2.01e9** (BT=40) | **all zeros, silent** |
| 192 | 2.42e9 (BT=48) | all zeros |
| 192 | 3.22e9 (BT=64) | all zeros |
| 120 | 2.01e9 (BT=64) | correct |
| 128 | 2.15e9 (BT=64) | correct |
| 136 | 2.28e9 (BT=64) | correct |
| 128 | 3.22e9 (BT=96) | correct |

So at C_in=192 the op breaks somewhere between 1.61e9 and 2.01e9 input elements,
while C_in=128 is still correct at 3.22e9 — the threshold is not a simple global
element or byte count, pointing at a specific kernel path (dispatch by channel
alignment?) with 32-bit-limited indexing.

For contrast, `mx.repeat` was verified correct producing 6.4e9-element outputs.

## Real-world impact

Found via mlx-video's Wan 2.1/2.2 VAE: its spatial upsampler folds video frames
into the batch dimension before a Conv2d (`[B*T, H, W, C]`), so decoding an
81-frame 768x512 video with default 64-frame temporal tiles crosses the threshold
at the `192 -> 96` upsample stage. Result: videos render as flat gray for exactly
the regions covered by large tiles, with no error anywhere — it presented as a
"tiling blending bug" and took a full layer-bisect plus per-op isolation to trace
here. Anyone composing large batched convs (video models, big batch inference)
can hit this and get silently wrong results.

Happy to run additional configurations on this machine if useful.

**Comment by jasonge27** [comment]
Thanks @kayadibi1 for reporting this. I ran into the same problem previously. 

Actually we can reproduce with a simpler case. On M4 Pro / 48GB (`max_recommended_working_set_size = 36GB`),

```
import mlx.core as mx, numpy as np

a = mx.full((3100, 1048576), 3.0)
b = mx.full((3100, 1048576), 4.0)
mx.eval(a, b)
c = a + b
mx.eval(c)
print(float(np.array(c[:1, :1])[0, 0]))   # 0.0
```
Here a, b, and c in total exceed command buffer, and driver kills the command buffer and MLX **drops the OOM error** silently.

I tried to instrumented commit() on HEAD (6a0dd0fd2):
```
[cb] status=4 signal_events=1 err=none
[cb] status=5 signal_events=0 err=Insufficient Memory (00000008:kIOGPUCommandBufferCallbackErrorOutOfMemory)
[cb] status=4 signal_events=1 err=none
3 + 4 = 0
[cb] DISCARDING unreported error: [METAL] Command buffer execution failed: ...
```
the failing buffer is the one `needs_commit()` commits inside `gpu::eval`.

Two questions @zcbenz :
1. Where should a command-buffer error live so it can't be lost — on the stream instead of the encoder, or attach the eval's synchronizer event to the mid-eval commit?
2. Worth a conv-side guard as well? max_unfold_rows() already caps explicit-gemm against maxBufferLength; winograd sizes inp_wg/out_wg from the full batch with nothing equivalent. That fixes this report but leaves the general case.

Happy to put up either :-)

**Comment by Gusanidas** [comment]
Opened #4102 for the conv2d side of this. It bounds Winograd’s working set by tiling the batch; the general silent command-buffer errors shown by the a + b repro are a separate issue.

**Comment by zcbenz** [comment]
> 1. Where should a command-buffer error live so it can't be lost — on the stream instead of the encoder, or attach the eval's synchronizer event to the mid-eval commit?

@jasonge27 Thanks for isolating the bug, I'm not sure about the answer, as I'm not sure where the OOM error was thrown and dropped, would have to do some experiments to know.

> 2. Worth a conv-side guard as well? max_unfold_rows() already caps explicit-gemm against maxBufferLength; winograd sizes inp_wg/out_wg from the full batch with nothing equivalent. That fixes this report but leaves the general case.

If the above fix is not enough I think a guard would help at least make the error verbose.

**Comment by jasonge27** [comment]
Thank you @zcbenz I'll submit a PR for the error handling and with the conv-side guard for your review.

**Comment by ternaus** [comment]
Additional 3D inference data point: this looks consistent with the general dropped-command-buffer-error path described above, not only the Conv2d Winograd manifestation.

Environment:

- MLX 0.32.0
- MacBook Pro, Apple M4 Max, 128 GB unified memory
- macOS 26.4.1 (25E253), Darwin 25.4.0
- Python 3.13

Workload: one FP16 3D nnU-Net-v1-style medical-segmentation network (one input channel, two output logits, 22 InstanceNorm3d layers). It receives one CT at a time. A normal training-context 128^3 patch agrees with PyTorch MPS FP16 at correlation 0.999998--0.999999; the issue appears only on much larger inference calls.

The first high-memory call was a two-tile inference over a CT with tiles 256 x 576 x 576. It returned a nonconstant result. Immediately after `mx.eval` completed, MLX reported 0.055 GiB active memory, 68.423 GiB cache memory, and 125.748 GiB peak memory. `mx.clear_cache()` changed the reported cache memory to 0, but the next CT inference caused the Python process to terminate without a Python exception.

In a previous long-running 68-CT run through the same model/checkpoint, two later full-volume calls saved both output channels as uint8 value 128 at every voxel (probability approximately 0.5 everywhere). The same raw CT preprocessing reproduces the stored FP16 model input byte-for-byte. Fresh isolated processes do not reproduce that all-128 result, which makes it unsafe to treat the persisted tensors as model predictions.

This is not a request to duplicate #4102. That PR addresses the Conv2d Winograd working set. The useful missing protection here is the general one discussed in this issue: a failed Metal command buffer must become a visible MLX exception instead of allowing a later `eval` to return/copy a plausible all-zero result. Is there a preferred way to add a focused Conv3d/large-graph regression or command-buffer error propagation test? I can narrow the reproducer further if that would help.

**Comment by zcbenz** [comment]
Closed by https://github.com/ml-explore/mlx/pull/4102
